### ISSUE TITLE
Failed to start QEMU binary /usr/libexec/qemu-kvm for probing: libvirt:  error : cannot execute binary /usr/libexec/qemu-kvm: Permission denied

### ISSUE THREAD

**Issue by unikzforce** [kubevirt/kubevirt#4303]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Hey there Guys, I had a problem with kind version 0.33 related to this [selinux issue](https://github.com/kubevirt/kubevirt/issues/4088), which I think the fix is not yet merged into release 0.33 yet, so I then decided to build kubevirt myself from latest commit (commit id 5029fafc) in master branch using [this guide](https://github.com/kubevirt/kubevirt/blob/master/docs/getting-started.md) while creating the kind cluster with a local registry with the help of [this guide](https://github.com/tilt-dev/kind-local).

And then restarted the procedure to start A CirrOS VM from [LAB 1](https://kubevirt.io/labs/kubernetes/lab1) But I faced another Issue:

this the result of the `kubectl describe vmi testvm`
```
+ kubectl describe vmi testvm
Name:         testvm
Namespace:    default
Labels:       kubevirt.io/domain=testvm
              kubevirt.io/nodeName=kind-control-plane
              kubevirt.io/size=small
Annotations:  kubevirt.io/latest-observed-api-version: v1alpha3
              kubevirt.io/storage-observed-api-version: v1alpha3
API Version:  kubevirt.io/v1alpha3
Kind:         VirtualMachineInstance
Metadata:
  Creation Timestamp:  2020-10-02T17:42:10Z
  Finalizers:
    foregroundDeleteVirtualMachine
  Generation:  6
  Managed Fields:
    API Version:  kubevirt.io/v1alpha3
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubevirt.io/latest-observed-api-version:
          f:kubevirt.io/storage-observed-api-version:
        f:labels:
          .:
          f:kubevirt.io/domain:
          f:kubevirt.io/nodeName:
          f:kubevirt.io/size:
        f:ownerReferences:
      f:spec:
        .:
        f:domain:
          .:
          f:devices:
            .:
            f:disks:
            f:interfaces:
          f:firmware:
            .:
            f:uuid:
          f:machine:
            .:
            f:type:
          f:resources:
            .:
            f:requests:
              .:
              f:memory:
        f:networks:
        f:volumes:
      f:status:
        .:
        f:activePods:
          .:
          f:14981c70-f492-4131-8e10-63fad752dc6e:
        f:guestOSInfo:
        f:nodeName:
        f:phase:
        f:qosClass:
    Manager:      virt-controller
    Operation:    Update
    Time:         2020-10-02T17:45:11Z
    API Version:  kubevirt.io/v1alpha3
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        f:conditions:
        f:migrationMethod:
    Manager:    virt-handler
    Operation:  Update
    Time:       2020-10-02T17:45:11Z
  Owner References:
    API Version:           kubevirt.io/v1alpha3
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  VirtualMachine
    Name:                  testvm
    UID:                   a8a1b7d3-412c-4e69-92f6-48a931c1b34a
  Resource Version:        3926
  Self Link:               /apis/kubevirt.io/v1alpha3/namespaces/default/virtualmachineinstances/testvm
  UID:                     5d474294-4a8c-46f1-a897-61dd60a274e9
Spec:
  Domain:
    Devices:
      Disks:
        Disk:
          Bus:  virtio
        Name:   containerdisk
        Disk:
          Bus:  virtio
        Name:   cloudinitdisk
      Interfaces:
        Bridge:
        Name:  default
    Features:
      Acpi:
        Enabled:  true
    Firmware:
      Uuid:  5a9fc181-957e-5c32-9e5a-2de5e9673531
    Machine:
      Type:  q35
    Resources:
      Requests:
        Cpu:     100m
        Memory:  64M
  Networks:
    Name:  default
    Pod:
  Volumes:
    Container Disk:
      Image:              kubevirt/cirros-registry-disk-demo
      Image Pull Policy:  Always
    Name:                 containerdisk
    Cloud Init No Cloud:
      userDataBase64:  SGkuXG4=
    Name:              cloudinitdisk
Status:
  Active Pods:
    14981c70-f492-4131-8e10-63fad752dc6e:  kind-control-plane
  Conditions:
    Last Probe Time:       <nil>
    Last Transition Time:  <nil>
    Message:               cannot migrate VMI which does not use masquerade to connect to the pod network
    Reason:                InterfaceNotLiveMigratable
    Status:                False
    Type:                  LiveMigratable
    Last Probe Time:       <nil>
    Last Transition Time:  2020-10-02T17:45:11Z
    Message:               server error. command SyncVMI failed: "LibvirtError(Code=1, Domain=10, Message='internal error: Failed to start QEMU binary /usr/libexec/qemu-kvm for probing: libvirt:  error : cannot execute binary /usr/libexec/qemu-kvm: Permission denied\n')"
    Reason:                Synchronizing with the Domain failed.
    Status:                False
    Type:                  Synchronized
  Guest OS Info:
  Migration Method:  BlockMigration
  Node Name:         kind-control-plane
  Phase:             Scheduled
  Qos Class:         Burstable
Events:
  Type     Reason            Age                From                              Message
  ----     ------            ----               ----                              -------
  Normal   SuccessfulCreate  3m13s              virtualmachine-controller         Created virtual machine pod virt-launcher-testvm-7hnrt
  Warning  SyncFailed        1s (x12 over 12s)  virt-handler, kind-control-plane  server error. command SyncVMI failed: "LibvirtError(Code=1, Domain=10, Message='internal error: Failed to start QEMU binary /usr/libexec/qemu-kvm for probing: libvirt:  error : cannot execute binary /usr/libexec/qemu-kvm: Permission denied\n')"
```

**What you expected to happen**:

I didn't expect `server error. command SyncVMI failed: "LibvirtError(Code=1, Domain=10, Message='internal error: Failed to start QEMU binary /usr/libexec/qemu-kvm for probing: libvirt:  error : cannot execute binary /usr/libexec/qemu-kvm: Permission denied\n')"` to happen

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`): master branch on commit id 5029fafc after release 0.33
- Kubernetes version (use `kubectl version`): 
```
+ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```
- VM or VMI specifications:
CirrOS on lab 1
- Cloud provider or hardware configuration:
core i7 + 16GB Mem
- OS (e.g. from /etc/os-release):
```
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```
- Kernel (e.g. `uname -a`):
`Linux unikz-ThinkPad-E550 5.4.0-48-generic #52-Ubuntu SMP Thu Sep 10 10:58:49 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
manually building from git.
- Others:

**Comment by nevonb** [comment]
I ran into this same issue. It looks like the virt-handler is trying to invoke the QEMU binary at /usr/libexec/qemu-kvm, which gets blocked by the AppArmor profile for libvirtd on Ubuntu-based systems. But more importantly, the qemu-kvm package on Ubuntu installs the binary with a different location and name (e.g., /usr/bin/qemu-system-x86_64). I was able to workaround it by creating a symlink for the binary and adding an entry for /usr/libexec/* to the usr.sbin.libvirtd profile in /etc/apparmor.d.

**Comment by unikzforce** [comment]
Thanks a lot, I'll Try It

**Comment by aodinokov** [comment]
Got the same with Ubuntu 18.04:
```
Linux aodinokov 4.15.0-122-generic #124-Ubuntu SMP Thu Oct 15 13:03:05 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
can see this in dmesg:
```
type=1400 audit(1608590689.669:246): apparmor="DENIED" operation="exec" profile="/usr/sbin/libvirtd" name="/usr/libexec/qemu-kvm" pid=142248 comm="rpc-worker" requested_mask="x" denied_mask="x" fsuid=107 ouid=0
```

the following on the host machines helped to workaround
Added `/usr/libexec/qemu-kvm PUx,` into /etc/apparmor.d/usr.sbin.libvirtd & and reloaded apparmor profiles: `sudo systemctl reload apparmor.service`

**Comment by moto-timo** [comment]
I see the same failure mechanism on Debian 10 "buster". Seems like there is an assumption in the code looking for /usr/libexec/qemu-kvm that is too rigid.

**Comment by unikzforce** [comment]
> Got the same with Ubuntu 18.04:
> 
> ```
> Linux aodinokov 4.15.0-122-generic #124-Ubuntu SMP Thu Oct 15 13:03:05 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
> ```
> 
> can see this in dmesg:
> 
> ```
> type=1400 audit(1608590689.669:246): apparmor="DENIED" operation="exec" profile="/usr/sbin/libvirtd" name="/usr/libexec/qemu-kvm" pid=142248 comm="rpc-worker" requested_mask="x" denied_mask="x" fsuid=107 ouid=0
> ```
> 
> the following on the host machines helped to workaround
> Added `/usr/libexec/qemu-kvm PUx,` into /etc/apparmor.d/usr.sbin.libvirtd & and reloaded apparmor profiles: `sudo systemctl reload apparmor.service`

Hey @moto-timo , Have you tried this? this worked for me well.

**Comment by moto-timo** [comment]
I didn't need to do that previously. There was zero issue running kubevirt a few weeks ago.

**Comment by moto-timo** [comment]
I finally got around to trying this again and it does work. I still don’t understand why it broke between 0.36/0.38 and 0.40

`$ sudo ln -s /usr/bin/kvm /usr/libexec/qemu-kvm`

And then the apparmor change above and restart the apparmor.service

**Comment by unikzforce** [comment]
@moto-timo So on debian kvm installation path is different

**Comment by francisco-pc7063** [comment]
virt-launcher pod still logs this out for me:
```
error: failed to get emulator capabilities
error: internal error: Failed to start QEMU binary /usr/libexec/qemu-kvm for probing: libvirt:  error : cannot execute binary /usr/libexec/qemu-kvm: Permission denied
```
Even after adding the symlink to /usr/libexec/qemu-kvm
```
$ ls -l /usr/libexec/qemu-kvm 
lrwxrwxrwx 1 root root 12 Sep 14 10:49 /usr/libexec/qemu-kvm -> /usr/bin/kvm
```
I'm not using apparmor.
Debian GNU/Linux 10 (buster)
I'm using minikube with docker driver

@moto-timo did you add any extra config?
